### PR TITLE
[ALLUXIO-2997] [S3] S3 multipart upload

### PR DESCRIPTION
**Pull request by uronce-cc**
https://alluxio.atlassian.net/browse/ALLUXIO-2997

This PR implements

[DONE] [multipart upload initialization](http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadInitiate.html)
[DONE] [upload part](http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html)
[DONE] [abort upload](http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadAbort.html)
[DONE] [complete upload](http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadComplete.html)
[DONE] [list parts](http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadListParts.html)

The following Python script initiates a multipart upload, uploads the parts, and completes the upload.

```python
import math, os

import boto
import boto.s3.connection
from filechunkio import FileChunkIO

# Connect to S3
c = boto.connect_s3(
    aws_access_key_id = '',
    aws_secret_access_key = '',
    host = 'localhost',
    port = 39999,
    path = '/api/v1/s3',
    is_secure=False,
    calling_format = boto.s3.connection.OrdinaryCallingFormat(),
)
b = c.create_bucket('mybucket')

# Get file info
source_path = '8mb.data'
source_size = os.stat(source_path).st_size

# Create a multipart upload request
mp = b.initiate_multipart_upload(os.path.basename(source_path))

# Use a chunk size of 1MB (feel free to change this)
chunk_size = 1048576
chunk_count = int(math.ceil(source_size / float(chunk_size)))

# Send the file parts, using FileChunkIO to create a file-like object
# that points to a certain byte range within the original file. We
# set bytes to never exceed the original file size.
for i in range(chunk_count):
    offset = chunk_size * i
    bytes = min(chunk_size, source_size - offset)
    with FileChunkIO(source_path, 'r', offset=offset, bytes=bytes) as fp:
        mp.upload_part_from_file(fp, part_num=i + 1)

# Finish the upload
mp.complete_upload()
```

After initializing the multipart upload:

```
./bin/alluxio fs ls /mybucket
drwxr-xr-x     changcheng     staff          8              09-01-2017 13:30:49:579  Directory      /mybucket/8mb.data_s3_multipart_tmp
```

After uploading the parts:

```
./bin/alluxio fs ls /mybucket/8mb.data_s3_multipart_tmp
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:621  In Memory      /mybucket/8mb.data_s3_multipart_tmp/1
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:687  In Memory      /mybucket/8mb.data_s3_multipart_tmp/2
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:741  In Memory      /mybucket/8mb.data_s3_multipart_tmp/3
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:779  In Memory      /mybucket/8mb.data_s3_multipart_tmp/4
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:820  In Memory      /mybucket/8mb.data_s3_multipart_tmp/5
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:853  In Memory      /mybucket/8mb.data_s3_multipart_tmp/6
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:892  In Memory      /mybucket/8mb.data_s3_multipart_tmp/7
-rw-r--r--     changcheng     staff          1048576        09-01-2017 13:30:49:926  In Memory      /mybucket/8mb.data_s3_multipart_tmp/8
```

After completing the upload:

```
./bin/alluxio fs ls /mybucket
-rw-r--r--     changcheng     staff          8388608        09-01-2017 13:30:35:854  Not In Memory  /mybucket/8mb.data
```

### CODE REVIEW

**Comment by uronce-cc** [comment]
alluxio-bot, check this please

**Comment by uronce-cc** [comment]
jenkins, test this please

**Comment by uronce-cc** [comment]
jenkins, test this please

**Comment by uronce-cc** [comment]
jenkins, test this please

**Comment by uronce-cc** [comment]
jenkins, test this please

File: core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+@JacksonXmlRootElement(localName = "CompleteMultipartUploadResult")
+@JsonPropertyOrder({ "Location", "Bucket", "Key", "ETag" })
+public class CompleteMultipartUploadResult {
+  // The URI that identifies the newly created object.
+  private String mLocation;
+  // Name of the bucket.
+  private String mBucket;
+  // Object key.
+  private String mKey;
+  // Entity tag of the object.
```
Comment (calvinjia) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java]: Could you change this to `/* */` format?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  /**
+   * @param storageClass the storage class to set
+   */
+  @JacksonXmlProperty(localName = "StorageClass")
+  public void setStorageClass(String storageClass) {
+    mStorageClass = storageClass;
+  }
+
+  /**
+   * @return whether (true) or not (false) all of the results were returned
```
Comment (calvinjia) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: Isn't it the other way around?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+   */
+  @JacksonXmlProperty(localName = "IsTruncated")
+  public void setIsTruncated(boolean isTruncated) {
+    mIsTruncated = isTruncated;
+  }
+
+  /**
+   * @return the list of parts
+   */
+  @JacksonXmlProperty(localName = "Part")
```
Comment (calvinjia) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: Should we make the localname `Parts`?

**Comment by uronce-cc** [comment]
@calvinjia ```Since there is a single success/failure returned, there is no need to support partial failures (ie. only a few of the parts made it).```, it's not true, if some parts are not uploaded correctly, then they can be uploaded again, if the follow up uploads succeeded, the parts can be completed into the final object.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java (hunk 3 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: the `Parts` are encoded into XML like:

```
<Part>
    ...
</Part>
<Part>
   ...
</Part>
```

File: core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java (hunk 1 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java (hunk 2 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 4]
```diff
@@ -257,6 +277,100 @@ public Response call() throws S3Exception {
...
+      public InitiateMultipartUploadResult call() throws S3Exception {
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        String objectPath = bucketPath + AlluxioURI.SEPARATOR + object;
+        AlluxioURI multipartTemporaryDir =
+            new AlluxioURI(S3RestUtils.getMultipartTemporaryDirForObject(bucketPath, object));
+
+        try {
+          mFileSystem.createDirectory(multipartTemporaryDir);
+          long uploadId = mFileSystem.getStatus(multipartTemporaryDir).getFileId();
```
Comment (madanadit) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: leave a comment about using fileId of temp dir as uploadId?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 5]
```diff
@@ -196,30 +199,47 @@ public ListBucketResult call() throws S3Exception {
...
+            + "missing together.");
 
         String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
         checkBucketIsAlluxioDirectory(bucketPath);
+
         String objectPath = bucketPath + AlluxioURI.SEPARATOR + object;
+        if (partNumber != null) {
+          // This object is part of a multipart upload, should be uploaded into the temporary
+          // directory first.
+          String tmpDir = S3RestUtils.getMultipartTemporaryDirForObject(bucketPath, object);
```
Comment (madanadit) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: maybe not for this PR, but is there a way for us to garbage collect abandoned multi-part uploads?. also please leave a comment if it makes sense

File: core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+   * @param location the URI that identifies the newly created object
+   * @param bucket name of the bucket
+   * @param key object key
+   * @param etag entity tag of the newly created object, the etag should not be surrounded by quotes
+   */
+  public CompleteMultipartUploadResult(String location, String bucket, String key, String etag) {
+    mLocation = location;
+    mBucket = bucket;
+    mKey = key;
+    mETag = "\"" + etag + "\"";
```
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java]: is there a way to check the etag param is not surrounded by quotes?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+  public void setParts(List<Part> parts) {
+    mParts = parts;
+  }
+
+  /**
+   * Part contains metadata of a part of the object in multipart upload.
+   */
+  @JsonPropertyOrder({ "PartNumber", "LastModified", "ETag", "Size" })
+  public static class Part {
+    // Part number.
```
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: nit: use `/* */`

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    public String getETag() {
+      return mETag;
+    }
+
+    /**
+     * @param etag the entity tag to set (without surrounding quotes)
+     */
+    @JacksonXmlProperty(localName = "ETag")
+    public void setETag(String etag) {
+      mETag = "\"" + etag + "\"";
```
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: extract `"\"" + etag + "\""` as a helper function in `S3RestUtils` ? it's used for a couple of times

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 5 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: according to the S3 specification, if the multipart upload is neither completed or aborted, the uploaded parts remain in the system.

Comment (madanadit) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: https://stackoverflow.com/questions/29911573/does-aws-s3-automatically-abort-multipart-uploads-after-a-timeout. Seems like may be able to configure a timeout.

File: core/common/src/main/java/alluxio/Constants.java [hunk 9]
```diff
@@ -153,6 +153,7 @@
   // S3 northbound API constants
   public static final String S3_DELETE_IN_ALLUXIO_ONLY = "ALLUXIO_ONLY";
   public static final String S3_DELETE_IN_ALLUXIO_AND_UFS = "ALLUXIO_AND_UFS";
+  public static final String S3_MULTIPART_TEMPORARY_DIR_SUFFIX = "_s3_multipart_tmp";
```
[the remark this answers is not in the document]
Comment (uronce-cc) [reply core/common/src/main/java/alluxio/Constants.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 4 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java (hunk 7 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java (hunk 8 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListPartsResult.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java (hunk 6 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadResult.java]: done

**Comment by uronce-cc** [comment]
@madanadit @ifcharming PTAL, thanks.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 5 above)
Comment (uronce-cc) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: thanks, left a TODO

**Pull request merged**